File: verification.go
package pact

import (
	"crypto/ed25519"
	"fmt"
	"net/http"
	"time"
)

// VerificationResult is the outcome of verifying a Pact-signed request.
type VerificationResult struct {
	// Valid is true if the request passed all verification checks.
	Valid bool `json:"valid"`

	// AgentID is the verified agent identity.
	AgentID string `json:"agent_id,omitempty"`

	// RootAuthority is the human (or root entity) who initiated the delegation chain.
	RootAuthority string `json:"root_authority,omitempty"`

	// Capabilities are the capabilities the agent is authorized to use.
	Capabilities []string `json:"capabilities,omitempty"`

	// ChainDepth is how many delegation links are in the chain.
	ChainDepth int `json:"chain_depth"`

	// Error describes why verification failed (if Valid is false).
	Error string `json:"error,omitempty"`

	// ExpiresAt is when the delegation chain expires (earliest expiry in chain).
	ExpiresAt time.Time `json:"expires_at,omitempty"`
}

// VerifyOptions configures the verification behavior.
type VerifyOptions struct {
	// MaxClockSkew is the maximum allowed difference between request timestamp
	// and server time. Default: 5 minutes.
	MaxClockSkew time.Duration

	// RequiredCapability is the capability the request must have.
	// If empty, only chain validity is checked (no capability authorization).
	RequiredCapability string

	// TrustedRoots is a set of known root authority IDs (typically human public key hashes).
	// If empty, any valid chain is accepted (open verification).
	// If set, the chain's root must be in this set.
	TrustedRoots map[string]bool

	// RevocationChecker is an optional function that checks if a delegation has been revoked.
	// Return true if the delegation ID has been revoked.
	RevocationChecker func(delegationID string) bool

	// Now overrides the current time (for testing).
	Now func() time.Time
}

// DefaultVerifyOptions returns sensible default verification options.
func DefaultVerifyOptions() VerifyOptions {
	return VerifyOptions{
		MaxClockSkew: 5 * time.Minute,
	}
}

// VerifyRequest performs full Pact verification on an HTTP request.
// This is the provider-side function — the main integration point.
//
// Verification steps:
//  1. Extract signature data from headers
//  2. Walk the delegation chain, verifying each link's signature
//  3. Check chain continuity (each link's "to" matches next link's "from")
//  4. Verify capabilities cover the required capability (if specified)
//  5. Check all constraints (expiry, chain depth, time bounds)
//  6. Verify the request signature against the terminal agent's public key
//  7. Check timestamp freshness (replay protection)
//
// All verification is LOCAL — zero network calls.
func VerifyRequest(req *http.Request, opts VerifyOptions) *VerificationResult {
	now := time.Now().UTC()
	if opts.Now != nil {
		now = opts.Now()
	}
	if opts.MaxClockSkew == 0 {
		opts.MaxClockSkew = 5 * time.Minute
	}

	// Step 1: Extract signature data
	sigData, err := ExtractSignatureData(req)
	if err != nil {
		return &VerificationResult{Valid: false, Error: fmt.Sprintf("extraction failed: %v", err)}
	}

	// Step 2-6: Verify chain + request
	return verifySignatureData(sigData, opts, now)
}

// VerifyChain verifies a delegation chain without an HTTP request.
// Useful for offline chain inspection and validation.
func VerifyChain(chain DelegationChain, opts VerifyOptions) *VerificationResult {
	now := time.Now().UTC()
	if opts.Now != nil {
		now = opts.Now()
	}

	return verifyChainOnly(chain, opts, now)
}

// verifySignatureData performs the full verification pipeline.
func verifySignatureData(sigData *RequestSignatureData, opts VerifyOptions, now time.Time) *VerificationResult {
	// Verify the delegation chain
	chainResult := verifyChainOnly(sigData.Chain, opts, now)
	if !chainResult.Valid {
		return chainResult
	}

	// Step 6: Verify request signature
	// The terminal agent's public key should match the claimed identity
	if sigData.AgentID != sigData.Chain.TerminalEntity() {
		return &VerificationResult{
			Valid: false,
			Error: fmt.Sprintf("agent identity %s does not match chain terminal %s",
				sigData.AgentID, sigData.Chain.TerminalEntity()),
		}
	}

	// Get the terminal agent's public key from the chain
	lastDelegation := sigData.Chain[len(sigData.Chain)-1]
	agentPubKeyBytes, err := decodeBase64URL(lastDelegation.To.PublicKey)
	if err != nil {
		return &VerificationResult{
			Valid: false,
			Error: fmt.Sprintf("invalid agent public key: %v", err),
		}
	}
	agentPubKey := ed25519.PublicKey(agentPubKeyBytes)

	// Verify the request signature
	if !ed25519.Verify(agentPubKey, []byte(sigData.SignatureBase), sigData.Signature) {
		return &VerificationResult{
			Valid: false,
			Error: "request signature verification failed",
		}
	}

	// Step 7: Check timestamp freshness (replay protection)
	if !sigData.Timestamp.IsZero() {
		skew := now.Sub(sigData.Timestamp)
		if skew < 0 {
			skew = -skew
		}
		if skew > opts.MaxClockSkew {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("request timestamp too far from server time (skew: %v, max: %v)",
					skew, opts.MaxClockSkew),
			}
		}
	}

	return chainResult
}

// verifyChainOnly verifies the delegation chain structure and signatures.
func verifyChainOnly(chain DelegationChain, opts VerifyOptions, now time.Time) *VerificationResult {
	if len(chain) == 0 {
		return &VerificationResult{Valid: false, Error: "empty delegation chain"}
	}

	var earliestExpiry time.Time
	var terminalCaps []string

	for i, d := range chain {
		// Verify delegation type and version
		if d.Type != "delegation" {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain link %d: invalid type %q", i, d.Type),
			}
		}

		// Verify chain continuity: each link's "to" must match next link's "from"
		if i > 0 {
			prev := chain[i-1]
			if prev.To.ID != d.From.ID || prev.To.PublicKey != d.From.PublicKey {
				return &VerificationResult{
					Valid: false,
					Error: fmt.Sprintf("chain link %d: discontinuity — previous 'to' (%s) != current 'from' (%s)",
						i, prev.To.ID, d.From.ID),
				}
			}
		}

		// Verify the delegator's signature
		fromPubKeyBytes, err := decodeBase64URL(d.From.PublicKey)
		if err != nil {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain link %d: invalid 'from' public key: %v", i, err),
			}
		}
		fromPubKey := ed25519.PublicKey(fromPubKeyBytes)

		// Verify identity matches public key
		expectedID := deriveID(fromPubKey)
		if expectedID != d.From.ID {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain link %d: 'from' ID does not match public key", i),
			}
		}

		// Reconstruct signing payload and verify signature
		payload, err := d.signingPayload()
		if err != nil {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain link %d: failed to reconstruct signing payload: %v", i, err),
			}
		}

		sigBytes, err := decodeBase64URL(d.Signature)
		if err != nil {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain link %d: invalid signature encoding: %v", i, err),
			}
		}

		if !ed25519.Verify(fromPubKey, payload, sigBytes) {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain link %d: signature verification failed", i),
			}
		}

		// Check expiry
		expires, err := time.Parse(time.RFC3339, d.Constraints.Expires)
		if err != nil {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain link %d: invalid expiry format: %v", i, err),
			}
		}
		if now.After(expires) {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain link %d: delegation expired at %s", i, d.Constraints.Expires),
			}
		}
		if earliestExpiry.IsZero() || expires.Before(earliestExpiry) {
			earliestExpiry = expires
		}

		// Check NotBefore
		if d.Constraints.NotBefore != "" {
			notBefore, err := time.Parse(time.RFC3339, d.Constraints.NotBefore)
			if err == nil && now.Before(notBefore) {
				return &VerificationResult{
					Valid: false,
					Error: fmt.Sprintf("chain link %d: delegation not yet valid (not before %s)", i, d.Constraints.NotBefore),
				}
			}
		}

		// Check chain depth constraint
		if i > 0 && len(chain) > chain[0].Constraints.MaxChainDepth {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain exceeds max depth %d", chain[0].Constraints.MaxChainDepth),
			}
		}

		// Verify narrowing: each link's capabilities must be covered by parent
		if i > 0 {
			parentCaps, err := ParseCapabilitySet(chain[i-1].Capabilities)
			if err != nil {
				return &VerificationResult{
					Valid: false,
					Error: fmt.Sprintf("chain link %d: invalid parent capabilities: %v", i-1, err),
				}
			}
			childCaps, err := ParseCapabilitySet(d.Capabilities)
			if err != nil {
				return &VerificationResult{
					Valid: false,
					Error: fmt.Sprintf("chain link %d: invalid capabilities: %v", i, err),
				}
			}
			if !parentCaps.Covers(childCaps) {
				return &VerificationResult{
					Valid: false,
					Error: fmt.Sprintf("chain link %d: capabilities exceed parent scope (narrowing violation)", i),
				}
			}
		}

		// Check revocation
		if opts.RevocationChecker != nil && opts.RevocationChecker(d.ID) {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("chain link %d: delegation %s has been revoked", i, d.ID),
			}
		}

		terminalCaps = d.Capabilities
	}

	// Check trusted roots
	rootID := chain.RootAuthority()
	if len(opts.TrustedRoots) > 0 && !opts.TrustedRoots[rootID] {
		return &VerificationResult{
			Valid: false,
			Error: fmt.Sprintf("root authority %s is not trusted", rootID),
		}
	}

	// Check required capability
	if opts.RequiredCapability != "" {
		termCaps, err := ParseCapabilitySet(terminalCaps)
		if err != nil {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("invalid terminal capabilities: %v", err),
			}
		}
		reqCap, err := ParseCapability(opts.RequiredCapability)
		if err != nil {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("invalid required capability: %v", err),
			}
		}
		if !termCaps.Covers(CapabilitySet{*reqCap}) {
			return &VerificationResult{
				Valid: false,
				Error: fmt.Sprintf("agent lacks required capability: %s", opts.RequiredCapability),
			}
		}
	}

	return &VerificationResult{
		Valid:         true,
		AgentID:       chain.TerminalEntity(),
		RootAuthority: rootID,
		Capabilities:  terminalCaps,
		ChainDepth:    len(chain),
		ExpiresAt:     earliestExpiry,
	}
}

File: sdk/python/pact/verification.py
"""Request and delegation chain verification for the Pact protocol.

All verification is LOCAL — zero network calls. A provider can verify
a Pact-signed request using only data present in the request headers.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Callable, Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from .capabilities import CapabilitySet, parse_capability, parse_capability_set
from .delegation import Delegation, DelegationChain, chain_root_authority, chain_terminal_entity
from .encoding import decode_base64url
from .identity import derive_id
from .signing import HttpRequest, extract_signature_data


@dataclass
class VerificationResult:
    """The outcome of verifying a Pact-signed request."""

    valid: bool = False
    """True if the request passed all verification checks."""

    agent_id: str = ""
    """The verified agent identity ID."""

    root_authority: str = ""
    """The root authority (human) who initiated the delegation chain."""

    capabilities: list[str] = field(default_factory=list)
    """The capabilities the agent is authorized to use."""

    chain_depth: int = 0
    """Number of delegation links in the chain."""

    error: str = ""
    """Describes why verification failed (if valid is False)."""

    expires_at: Optional[datetime] = None
    """When the delegation chain expires (earliest expiry)."""


@dataclass
class VerifyOptions:
    """Configuration for verification behavior."""

    max_clock_skew: timedelta = field(
        default_factory=lambda: timedelta(minutes=5))
    """Maximum allowed difference between request timestamp and server time."""

    required_capability: str = ""
    """Capability the request must have. If empty, only chain validity is checked."""

    trusted_roots: set[str] = field(default_factory=set)
    """Known root authority IDs. If empty, any valid chain is accepted."""

    revocation_checker: Optional[Callable[[str], bool]] = None
    """Function that checks if a delegation ID has been revoked."""

    now: Optional[Callable[[], datetime]] = None
    """Override current time (for testing)."""


def default_verify_options() -> VerifyOptions:
    """Return sensible default verification options."""
    return VerifyOptions()


def verify_request(req: HttpRequest, opts: VerifyOptions | None = None) -> VerificationResult:
    """Perform full Pact verification on an HTTP request.

    This is the provider-side function — the main integration point.

    Verification steps:
    1. Extract signature data from headers
    2. Walk the delegation chain, verifying each link's signature
    3. Check chain continuity
    4. Verify capabilities cover the required capability (if specified)
    5. Check all constraints (expiry, chain depth, time bounds)
    6. Verify the request signature against the terminal agent's public key
    7. Check timestamp freshness (replay protection)

    All verification is LOCAL — zero network calls.
    """
    if opts is None:
        opts = default_verify_options()

    now = datetime.now(timezone.utc)
    if opts.now is not None:
        now = opts.now()

    if opts.max_clock_skew == timedelta(0):
        opts.max_clock_skew = timedelta(minutes=5)

    # Step 1: Extract signature data
    try:
        sig_data = extract_signature_data(req)
    except ValueError as e:
        return VerificationResult(valid=False, error=f"extraction failed: {e}")

    # Step 2-6: Verify chain + request
    return _verify_signature_data(sig_data, opts, now)


def verify_chain(chain: DelegationChain, opts: VerifyOptions | None = None) -> VerificationResult:
    """Verify a delegation chain without an HTTP request.

    Useful for offline chain inspection and validation.
    """
    if opts is None:
        opts = default_verify_options()

    now = datetime.now(timezone.utc)
    if opts.now is not None:
        now = opts.now()

    return _verify_chain_only(chain, opts, now)


def _verify_signature_data(
    sig_data: dict, opts: VerifyOptions, now: datetime
) -> VerificationResult:
    """Perform the full verification pipeline."""
    chain: DelegationChain = sig_data["chain"]

    # Verify the delegation chain
    chain_result = _verify_chain_only(chain, opts, now)
    if not chain_result.valid:
        return chain_result

    # Step 6: Verify request signature
    if sig_data["agent_id"] != chain_terminal_entity(chain):
        return VerificationResult(
            valid=False,
            error=f"agent identity {sig_data['agent_id']} does not match chain terminal {chain_terminal_entity(chain)}",
        )

    # Get the terminal agent's public key
    last = chain[-1]
    try:
        agent_pub_bytes = decode_base64url(last.to_entity.public_key)
        agent_pub_key = Ed25519PublicKey.from_public_bytes(agent_pub_bytes)
    except Exception as e:
        return VerificationResult(valid=False, error=f"invalid agent public key: {e}")

    # Verify the request signature
    try:
        agent_pub_key.verify(
            sig_data["signature"],
            sig_data["signature_base"].encode("utf-8"),
        )
    except Exception:
        return VerificationResult(
            valid=False, error="request signature verification failed"
        )

    # Step 7: Check timestamp freshness
    timestamp = sig_data.get("timestamp")
    if timestamp is not None:
        if timestamp.tzinfo is None:
            timestamp = timestamp.replace(tzinfo=timezone.utc)
        skew = abs((now - timestamp).total_seconds())
        if skew > opts.max_clock_skew.total_seconds():
            return VerificationResult(
                valid=False,
                error=f"request timestamp too far from server time (skew: {skew:.0f}s, max: {opts.max_clock_skew.total_seconds():.0f}s)",
            )

    return chain_result


def _verify_chain_only(
    chain: DelegationChain, opts: VerifyOptions, now: datetime
) -> VerificationResult:
    """Verify the delegation chain structure and signatures."""
    if not chain:
        return VerificationResult(valid=False, error="empty delegation chain")

    earliest_expiry: Optional[datetime] = None
    terminal_caps: list[str] = []

    for i, d in enumerate(chain):
        # Verify delegation type
        if d.type != "delegation":
            return VerificationResult(
                valid=False, error=f"chain link {i}: invalid type {d.type!r}"
            )

        # Verify chain continuity
        if i > 0:
            prev = chain[i - 1]
            if prev.to_entity.id != d.from_entity.id or prev.to_entity.public_key != d.from_entity.public_key:
                return VerificationResult(
                    valid=False,
                    error=f"chain link {i}: discontinuity — previous 'to' ({prev.to_entity.id}) != current 'from' ({d.from_entity.id})",
                )

        # Verify the delegator's signature
        try:
            from_pub_bytes = decode_base64url(d.from_entity.public_key)
            from_pub_key = Ed25519PublicKey.from_public_bytes(from_pub_bytes)
        except Exception as e:
            return VerificationResult(
                valid=False, error=f"chain link {i}: invalid 'from' public key: {e}"
            )

        # Verify identity matches public key
        expected_id = derive_id(from_pub_bytes)
        if expected_id != d.from_entity.id:
            return VerificationResult(
                valid=False,
                error=f"chain link {i}: 'from' ID does not match public key",
            )

        # Reconstruct signing payload and verify signature
        payload = d.signing_payload()
        try:
            sig_bytes = decode_base64url(d.signature)
        except Exception as e:
            return VerificationResult(
                valid=False,
                error=f"chain link {i}: invalid signature encoding: {e}",
            )

        try:
            from_pub_key.verify(sig_bytes, payload)
        except Exception:
            return VerificationResult(
                valid=False,
                error=f"chain link {i}: signature verification failed",
            )

        # Check expiry
        try:
            expires = datetime.fromisoformat(d.constraints.expires)
            if expires.tzinfo is None:
                expires = expires.replace(tzinfo=timezone.utc)
        except Exception as e:
            return VerificationResult(
                valid=False,
                error=f"chain link {i}: invalid expiry format: {e}",
            )
        if now > expires:
            return VerificationResult(
                valid=False,
                error=f"chain link {i}: delegation expired at {d.constraints.expires}",
            )
        if earliest_expiry is None or expires < earliest_expiry:
            earliest_expiry = expires

        # Check NotBefore
        if d.constraints.not_before:
            try:
                not_before = datetime.fromisoformat(d.constraints.not_before)
                if not_before.tzinfo is None:
                    not_before = not_before.replace(tzinfo=timezone.utc)
                if now < not_before:
                    return VerificationResult(
                        valid=False,
                        error=f"chain link {i}: delegation not yet valid (not before {d.constraints.not_before})",
                    )
            except Exception:
                pass

        # Check chain depth
        if i > 0 and len(chain) > chain[0].constraints.max_chain_depth:
            return VerificationResult(
                valid=False,
                error=f"chain exceeds max depth {chain[0].constraints.max_chain_depth}",
            )

        # Verify narrowing
        if i > 0:
            try:
                parent_caps = parse_capability_set(chain[i - 1].capabilities)
                child_caps = parse_capability_set(d.capabilities)
                if not parent_caps.covers(child_caps):
                    return VerificationResult(
                        valid=False,
                        error=f"chain link {i}: capabilities exceed parent scope (narrowing violation)",
                    )
            except ValueError as e:
                return VerificationResult(
                    valid=False,
                    error=f"chain link {i}: invalid capabilities: {e}",
                )

        # Check revocation
        if opts.revocation_checker is not None and opts.revocation_checker(d.id):
            return VerificationResult(
                valid=False,
                error=f"chain link {i}: delegation {d.id} has been revoked",
            )

        terminal_caps = d.capabilities

    # Check trusted roots
    root_id = chain_root_authority(chain)
    if opts.trusted_roots and root_id not in opts.trusted_roots:
        return VerificationResult(
            valid=False, error=f"root authority {root_id} is not trusted"
        )

    # Check required capability
    if opts.required_capability:
        try:
            term_caps = parse_capability_set(terminal_caps)
            req_cap = parse_capability(opts.required_capability)
            from .capabilities import CapabilitySet
            if not term_caps.covers(CapabilitySet([req_cap])):
                return VerificationResult(
                    valid=False,
                    error=f"agent lacks required capability: {opts.required_capability}",
                )
        except ValueError as e:
            return VerificationResult(
                valid=False, error=f"invalid capability: {e}"
            )

    return VerificationResult(
        valid=True,
        agent_id=chain_terminal_entity(chain),
        root_authority=root_id,
        capabilities=terminal_caps,
        chain_depth=len(chain),
        expires_at=earliest_expiry,
    )
